### ISSUE TITLE
Redirect URL https://outlook.office365.com/owa/ is valid?

### ISSUE THREAD

**Issue by itzamnamx** [O365/python-o365#187]
I got this simple code

from O365 import Account
import logging
logging.basicConfig(level=logging.DEBUG)

credentials = ('XXXXX', 'YYYYYYY')

account = Account(credentials=(credentials))

result = account.authenticate(scopes=['basic',   'message_all'])

I run it

$python3 simple.py 

DEBUG:requests_oauthlib.oauth2_session:Generated new state wKN3BwWj16LVX5iWLZ2K77YlPqHp5S.
Visit the following url to give consent:
https://login.microsoftonline.com/common/oauth2/v2.0/authorize?response_type=code&client_id=XXXXXX&redirect_uri=https%3A%2F%2Foutlook.office365.com%2Fowa%2F&scope=https%3A%2F%2Fgraph.microsoft.com%2FUser.Read+https%3A%2F%2Fgraph.microsoft.com%2FMail.ReadWrite+offline_access+https%3A%2F%2Fgraph.microsoft.com%2FMail.Send&state=wKN3BwWj16LVX5iWLZ2K77YlPqHp5S&access_type=offline
Paste the authenticated url here: 

I got "https://outlook.office365.com/owa/" as redirect URL also I got a permissions window and a login window, but what is the url that I should paste into the command line as input?.

**Comment by janscas** [comment]
After giving consent you will be redirected to outlook owa. The url ahould contain a query param called code. Copy the whole url and paste it back to account.authenticate

**Comment by roycem90** [comment]
If the url you are redirected to is simply https://outlook.office.com/owa/?realm=blahblah, and nothing else after that.. then you are currently on new Outlook look, revert back to old look and try the authentication flow again

**Comment by itzamnamx** [comment]
Let me explain better the issue that I am facing, trying to explain step by step:

1. I have this simple program with this content:

from O365 import Account, FileSystemTokenBackend, oauth_authentication_flow
import datetime as dt
import logging

logging.basicConfig()
logger = logging.getLogger()
logger.setLevel(logging.DEBUG)


credentials = ('XXXXXXX-f06b-465f-9dc1-XXXXXXX', 'XXXXXXXXX=$:%')

scopes = ['message_all']

account = Account(credentials)

if not account.is_authenticated:  # will check if there is a token and has not expired
    # ask for a login
    account.authenticate(scopes=scopes)

2. I run the program

$ python3 simple.py 
DEBUG:requests_oauthlib.oauth2_session:Generated new state eTXJPsRMxcsWBh3tTY8y9oh3gXXXX.
Visit the following url to give consent:
https://login.microsoftonline.com/common/oauth2/v2.0/authorize?response_type=code&client_id=XXXXXXXX&redirect_uri=https%3A%2F%2Foutlook.office365.com%2Fowa%2F&scope=https%3A%2F%2Fgraph.microsoft.com%2FMail.ReadWrite+https%3A%2F%2Fgraph.microsoft.com%2FMail.Send&state=eTXJPsRMxcsWBh3tTY8y9oh3gXXXX&access_type=offline
Paste the authenticated url here: 

3. I open the URL and I got the following page (I use the credentials of another account)

![seleccion_067](https://user-images.githubusercontent.com/10892046/52372622-4108ec00-2a1e-11e9-8344-9664bed9ff56.png)

4. I clicked on "YES" and then I got this page

![seleccion_068](https://user-images.githubusercontent.com/10892046/52372613-3c443800-2a1e-11e9-8dc2-43490acb2427.png)

5. If I paste the URL that is in the browser I got this message

Paste the authenticated url here: https://login.microsoftonline.com/common/oauth2/authorize?client_id=00000002-0000-0ff1-ce00-000000000000&redirect_uri=https%3a%2f%2foutlook.office365.com%2fowa%2f&resource=00000002-0000-0ff1-ce00-000000000000&response_mode=form_post&response_type=code+id_token&scope=openid&msafed=0&client-request-id=ef64f3b7-e912-4b50-a30c-0956eXXXXXX&protectedtoken=true&nonce=636850824992438019.ca42276f-256c-4e26-ad57-f75XXXXXXX&state=Dcu9CsIwFEDhVN_FLTa_NzdDEcRJKIhU1DG5SdutYivVtzfD-bZTMca2pU2pEgXmQANagcp4r4xGIf2eglHKQc-VBeImK-AhWcd7ZwlETBilrMp7q6c11AeaUm5acC5gjoYHLJCXPdfSZo5JyChN1hRoNy9hyU3uHufLfG2_NN-Po166J_78NOrhXXXXXX
ERROR:O365.connection:Unable to fetch auth token. Error: (missing_code) Missing code parameter in response.
Something go wrong. Please try again.

6. Stuck here

I can see that I have granted the application, because I receive the email of confirmation and also the information is displayed in my account about the grants that have the application, but I am unable to see where is the token or how can I recover the token in order to do the calls for creating mails.

Any help is very valuable.

Regards

Victor

**Comment by janscas** [comment]
> 4. I clicked on "YES" and then I got this page (again it request to sign in)

After this did you login?

You should be redirected to the outlook owa inbox
Then immediately copy that url. It should contain a "&code=xxxxx" query param.

The url that you are pasting back does not have the code query param.

**Comment by janscas** [comment]
> scopes = ['message_all']

Also note that if you don't include the 'basic' scope you will need to re-authenticate every hour. Include it so you don't have.

**Comment by itzamnamx** [comment]
I have tried with these code

from O365 import Account, FileSystemTokenBackend, oauth_authentication_flow
import datetime as dt
import logging

logging.basicConfig()
logger = logging.getLogger()
logger.setLevel(logging.DEBUG)


credentials = ('56530a20-XXXXXXX', 'XXXXXXX=$:%')

scopes = ['basic','message_all']

account = Account(credentials)

if not account.is_authenticated:  # will check if there is a token and has not expired
    # ask for a login
    account.authenticate(scopes=scopes)

And same behaviour, both accounts are new, created two days ago. What I have notice is that now a message is suggesting to manage the application from Azure and there I see this:

![image](https://user-images.githubusercontent.com/10892046/52376006-b082d980-2a26-11e9-88cd-6d06b94c15de.png)

I will try to enable and test it.

**Comment by itzamnamx** [comment]
Unfortunately same issue.

**Comment by janscas** [comment]
I've noticed you uncheck 'Allow implicit Flow' when registering the app in the platform section.

Please, check it and try again.

**Comment by gladiusunus** [comment]
I've got the same issue. spent all of today and several hours last week trying to get python to send emails. At my old job I had a simple process of passing username and password and it would send emails but that doesn't appear to work anymore.

**Comment by janscas** [comment]
Microsoft deprecated the basic auth (username and password) on nov last year.

Make sure you are NOT using the new outlook owa interface.

Maybe if you just put some code and logs I can help more...

**Comment by gladiusunus** [comment]
1. py code used to generate authentication url
2. auth url and request for authenticated url
3. accept permissions 
4. then the redirected page just shows the basic owa url. 

I saw in this thread that the '*code=*' url is what is wanted but i never saw anything with it. 

(blacked out email address, client id, and client secret)



![py code 1](https://user-images.githubusercontent.com/47797462/53051408-6efc2080-3461-11e9-9530-e567f0473e47.PNG)
![url 1](https://user-images.githubusercontent.com/47797462/53051409-6efc2080-3461-11e9-87e2-5d34aab50648.PNG)
![permissions](https://user-images.githubusercontent.com/47797462/53051406-6efc2080-3461-11e9-8faa-e6417f680b58.PNG)
![url2](https://user-images.githubusercontent.com/47797462/53051410-6efc2080-3461-11e9-8408-460c545a7c4c.PNG)

**Comment by janscas** [comment]
@gladiusunus you are passing incorrect formatted scopes:

Acording to your screenshots are doing this:

```python
scopes = ['basic', 'message_all'] # this is equal to pass: ['offline_access', 'https://graph.microsoft.com/User.Read', 'https://graph.microsoft.com/Mail.ReadWrite', 'https://graph.microsoft.com/Mail.Send']
credentials = (client_id, client_secret)
account = Account(credentials)
scopes = ['Mail.ReadWrite', 'Mail.Send', 'offline_access', '...'] # BAD SCOPES!!

# you can use scope helpers (the first scopes line) or scopes prefixed with the protocol url (https://graph.microsoft.com/)

account.authenticate(scopes=scopes)  # you are passing the bad scopes here.
```

Instead, use scope helpers or prefix all your scopes with 'https://graph.microsoft.com/' (all except 'offline_access').


Try again with the correct scopes and using the old outlook owa interface.

**Comment by gladiusunus** [comment]
@janscas, I tried that with 
`scopes = ['https://graph.microsoft.com/Mail.ReadWrite', 'https://graph.microsoft.com/Mail.Send']`
and 
`scopes = ['basic', 'message_all']`

and same thing. it give me a screen to confirm my account then the https://outlook.office365.com then i follow the link to my mailbox online. still no url containing 'code'.

to be clear, when i past whatever url code i try i get the error message 
`Unable to fetch auth token. Error: (insecure_transport) OAuth 2 MUST utilize https.` 
is that because the url doesn't contain the code it's looking for?

too i've been told my company IT dept (in a different convo about similar topic) that i was trying to use the O365 api and they told me that it could be a lack of permissioning. is there a permission check i can do to see if this is even worth me spending more time on?

**Comment by janscas** [comment]
You aren’t supose to follow any link. The link is opened automatically. After giving login/consent you will be redirected automatically to outlook owa.

And the error you are receiving is because you are doing something wrong. The url must begin with https.

**Comment by gladiusunus** [comment]
ahh, i was pasting the authenticated url in qoutes.
`'https://blahblahblah'`

so when i don't do that the error message changes to 
`Unable to fetch auth token. Error: (missing_code) Missing code parameter in response.`

so the auth token should be in the url correct? followed by something along the lines of 'code='? i haven't seen a url with code in it yet besides the url to give consent ('respose_type=code&') which i know isn't what i'm looking for. 

is there permission setting that i can check to confirm that IT hasn't removed to prevent people from using the api?

**Comment by janscas** [comment]
> followed by something along the lines of 'code='

yep you later exchanges this 'code' with an auth token.

I don't really know why. It does not happen to me.

**Comment by rampeer** [comment]
Ran into the same issue.

I've noticed that I cannot log in to
>https://outlook.office365.com 
at all. It says that I should not use a personal account, and use "work account" instead. I could not find much information on how to obtain the latter, and what are the differences.

I can access OWA using 
>https://outlook.live.com/mail/inbox
However, I cannot use it as a redirect URL.

Sometimes, I feel like Microsoft just does not want to cooperate.

**Comment by coldbyte** [comment]
Got this problem if i activate the new outlook:

![image](https://user-images.githubusercontent.com/1881581/54828894-1cef2a80-4cad-11e9-9a41-79f69cf64d70.png)

**Comment by lucasrodrigues10** [comment]
Me too @coldbyte , but after disabling it worked.

**Comment by rampeer** [comment]
Where can I find this flag? I could not find it in my OWA or desktop application. I tried searching for it, but the only relevant response is 

https://answers.microsoft.com/en-us/outlook_com/forum/all/how-to-switch-back-to-outlook-classic/278603c6-c863-4080-a77e-f7324afcdd40?auth=1

and it states that all users were forcibly switched to new versions, and there is no way to switch back.

Would you kindly provide a slightly more detailed description? It would help me - and many other people who will stumble upon this issue.

**Comment by janscas** [comment]
> I've noticed that I cannot log in to https://outlook.office365.com at all. It says that I should not use a personal account, and use "work account" instead. I could not find much information on how to obtain the latter, and what are the differences.
> 
> I can access OWA using https://outlook.live.com/mail/inbox . However, I cannot use it as a redirect URL, as after log-in I get redirected to an error page stating that redirect URL is not valid.

that's odd

**Comment by nnit-kzhg** [comment]
i got this exactly same issue. I will redirect to my org https://owa webpage, after login, there is just a url without any code, the URL just like below, could somebody fix it?

https://owa.nnit.com/owa/'myname@mydomain.com'/#path=/mail

**Comment by rampeer** [comment]
One alternative is using Exchange protocol if your server supports it (using https://github.com/ecederstrand/exchangelib for example ). I ended up using it for Office 365 calendar manipulations.

**Comment by janscas** [comment]
Hi guys. I have the solution. I'm posting it here and also fixing the whole authentication urls for the new version released later today (1.1.4)

**Comment by janscas** [comment]
Well ... this one was quite difficult to figure it out.

You have to follow this two points:

1. Use the [azure portal](https://portal.azure.com/) to register applications. Go to the "Application Registrations" section and **switch to the preliminary version**(important). The old [App Registration Portal](https://apps.dev.microsoft.com) will disappear in May 2019. If you go there you already will be suggested to use the azure portal (with the preliminary version switched on.
This also seems to fix the problems when you are using a personal account. I haven't test this (I can't as I don't have a personal account), but unlike the old app registration portal, in azure you can set the audience to be any outlook account (personal, work, etc). I will update the readme with the steps needed to change this.

1. Prior to version 1.1.4 you will have to pass to `account.authenticate` this param: `redirect_uri` with this value:  "https://login.microsoftonline.com/common/oauth2/nativeclient". This redirect uri seems to work no matter what. At least it does to me and it is documented by Microsoft to work for this cases. You also have to set this redirect uri when registering the app.


**TL;DR:**

First of all, using "https://outlook.office365.com/owa/" as the redirect uri is odd. This url was never intended to work as a redirect uri. It worked but not now with the new interface (which will be the default sometime). 
Fortunately you can change this redirect uri in the O365 library easily. In fact if you have your own website and you are using https, then you can redirect there and grab the code. But for those not having this option you have to conform with "https://outlook.office365.com/owa/" and now with the much better "https://login.microsoftonline.com/common/oauth2/nativeclient".

This seems to solve all the "code not found" problems.

For the other kind of problems, it seems that using a personal account vs a work account have it's issues as well. I couldn't find a solution here. And I'm not sure if this can solve the problem, but I see that you can change the target audience in the azure portal from work accounts to any microsoft account (personal, educational, etc.).
I will be glad if someone can set some light on this matter.

Please report back if this solved your issues.

**Comment by janscas** [comment]
> For the other kind of problems, it seems that using a personal account vs a work account have it's issues as well. I couldn't find a solution here. And I'm not sure if this can solve the problem, but I see that you can change the target audience in the azure portal from work accounts to any microsoft account (personal, educational, etc.).
> I will be glad if someone can set some light on this matter.

I've just tried with a personal account (https://outlook.live.com/) and it worked.

**Comment by janscas** [comment]
version 1.1.4 now ready on pypi

**Comment by thomas-schulze-entwicklerschmiede** [comment]
@janscas Hi Alejandro,
I tried the current version without any success. The older version with this peculiar /owa redirect URL worked. Although I get back an URL with a code-parameter, using that URL gives:

> O365.connection[14254] ERROR Unable to fetch auth token. Error: (invalid_client) xxxx: Client is public so 'client_assertion' should not be presented.

Perhaps You can suggest, what I am missing here!?

Many thanks in advance!

**Comment by janscas** [comment]
Hi,did you register the app as a “web app”?

**Comment by thomas-schulze-entwicklerschmiede** [comment]
[Bildschirmfoto 2019-04-20 um 10.50.08.pdf](https://github.com/O365/python-o365/files/3099901/Bildschirmfoto.2019-04-20.um.10.50.08.pdf)

this is my configuration!?

**Comment by thomas-schulze-entwicklerschmiede** [comment]
Did I miss something?

**Comment by janscas** [comment]
@thomas-schulze-entwicklerschmiede i have it like this:
![056F4AE3-4318-4B34-B6A4-26B2123A2DF0](https://user-images.githubusercontent.com/7224106/56455391-ef55d980-635d-11e9-8330-a7486d00869c.png)

**Comment by janscas** [comment]
Go to the app manifest and check that you have this:

```
"replyUrlsWithType": [
		{
			"url": "https://login.microsoftonline.com/common/oauth2/nativeclient",
			"type": "Web"
		}
	],
```

**Comment by thomas-schulze-entwicklerschmiede** [comment]
@janscas Hi Alejandro,

thanks a lot - that was indeed the issue!

Many thanks!

**Comment by sbi-rviot** [comment]
> If the url you are redirected to is simply https://outlook.office.com/owa/?realm=blahblah, and nothing else after that.. then you are currently on new Outlook look, revert back to old look and try the authentication flow again

Thank you this helped me … But I do not understand the reason. Would you have one?

**Comment by gazcn007** [comment]
I am running into the issue where after you copy and paste the oauth2 url, the python process just gets stuck:
1. in my code
```
account = Account(credentials, scopes=scopes)
account.authenticate()
```
2. i run the python program in the terminal, it prints
```
Visit the following url to give consent:
https://login.microsoftonline.com/common/oauth2/v2.0/authorize?response_type=code&client_id=xxx&redirect_uri=xxx&scope=xxxxx&access_type=offline
```
3. I go to that URL, authenticate no problem, it redirects me to an empty `https://login.microsoftonline.com/common/oauth2/nativeclient` link with `code` parameter
4. I paste it back to the terminal:
```
Paste the authenticated url here:
https://login.microsoftonline.com/common/oauth2/nativeclient?code=OAQABAAIAAABeAFzDwllzTYGDLh_qYbH8Vb7ZOYsqE8Xhau8b0UleSsosHLQXb2PDQPBmbfk9P_u_aIlCRbyXaDL_53dgmrh3hk534HeTQZpEHnAbz38z7__sNiDtsqi9z0ZQpgPacpGzIfLcQQ1sZDTSMInM_-lZQ93o9jQbdVs23-syGYDZS-J4ODrrzZAtX8LB58u1U3BnfHBOqcRKRmBEAAAuSyReIEKGLzEINNLeqtnsVCCqI2bv77OldpxS7cngdDnq4CV3kVTAH743ls3dOPQxkbyWCdex_lxpyRgn2DS9rGvg4wUygV8LOJdbPrNeBQDpbS1bEPPAFebXKPz_zrdagElZW2zOLjWLS4pMONE4wNZ1LpekSoEi1tiXi-y3tz1X42C8KesB33sX7FOsriaDcjXU9BFRg4xBwHxIonHffezBYePj9BT0EePu5x2iOVbJLhGJjUyFGTMSgjTzXoJKA3DaIxRcDiY8c4W1WA-AA&state=ZX7ivCbDUKWaqc6VheU
```

And it will just be stuck forever.

I have configured:

```
"replyUrlsWithType": [
		{
			"url": "https://login.microsoftonline.com/common/oauth2/nativeclient",
			"type": "Web"
		}
	],
```

**Comment by janscas** [comment]
When pasting into the console and depending on your used OS, terminal, etc.. It can be tricky to correctly paste the value and continue the flow of the script.

I had problems also with using the **cmd** in windows. In this case you just have to paste the string and hit enter moving the cursor to another line. If you hit the "enter" right after pasting it get's stuck.

Also check this Tip from the readme:

> Tip: When using MacOs the console is limited to 1024 characters. If your url has multiple scopes it can exceed this limit. To solve this. Just import readline a the top of your script.

**Comment by zoomhgtk** [comment]
Dear Janscas,

Could you please share more details about how I should solve the issue of MacOs console limited to 1024 characters. 
I am quite confused about how to use readline to get the pasted URL to the interactive python function.

Thanks.
